web_src/js/features/repo-view-file-tree-sidebar.ts

```diff
@@ -0,0 +1,103 @@
+import {createApp, ref} from 'vue';
+import {toggleElem} from '../utils/dom.ts';
+import {GET, PUT} from '../modules/fetch.ts';
+import {registerGlobalInitFunc} from '../modules/observer.ts';
+import ViewFileTree from '../components/ViewFileTree.vue';
+
+async function toggleSidebar(sidebarEl: HTMLElement, visibility: boolean) {
+  const showBtnEl = sidebarEl.parentElement.querySelector('.show-tree-sidebar-button');
+  const containerClassList = sidebarEl.parentElement.classList;
+  containerClassList.toggle('repo-grid-tree-sidebar', visibility);
+  containerClassList.toggle('repo-grid-filelist-only', !visibility);
+  toggleElem(sidebarEl, visibility);
+  toggleElem(showBtnEl, !visibility);
+
+  // FIXME: need to remove "full height" style from parent element
+
+  if (!sidebarEl.hasAttribute('data-is-signed')) return;
+
+  // save to session
+  await PUT('/repo/preferences', {
+    data: {
+      show_file_view_tree_sidebar: visibility,
+    },
+  });
+}
+
+function childrenLoader(sidebarEl: HTMLElement) {
+  return async (path: string, recursive?: boolean) => {
+    const fileTree = sidebarEl.querySelector('#view-file-tree');
+    const apiBaseUrl = fileTree.getAttribute('data-api-base-url');
+    const refTypeNameSubURL = fileTree.getAttribute('data-current-ref-type-name-sub-url');
+    const response = await GET(encodeURI(`${apiBaseUrl}/tree/${refTypeNameSubURL}/${path ?? ''}?recursive=${recursive ?? false}`));
+    const json = await response.json();
+    if (json instanceof Array) {
+      return json.map((i) => ({
+        name: i.name,
+        type: i.type,
+        path: i.path,
+        sub_module_url: i.sub_module_url,
+        children: i.children,
+      }));
+    }
+    return null;
+  };
+}
+
+async function loadContent(sidebarEl: HTMLElement) {
+  // load content by path (content based on home_content.tmpl)
+  const response = await GET(`${window.location.href}?only_content=true`);
+  const contentEl = sidebarEl.parentElement.querySelector('.repo-home-filelist');
+  contentEl.innerHTML = await response.text();
+  reloadContentScript(sidebarEl, contentEl);
+}
+
+function reloadContentScript(sidebarEl: HTMLElement, contentEl: Element) {
+  contentEl.querySelector('.show-tree-sidebar-button')?.addEventListener('click', () => {
+    toggleSidebar(sidebarEl, true);
+  });
+}
+
+export function initViewFileTreeSidebar() {
+  let popstateListenerForViewFilePageAdded = false;
+  registerGlobalInitFunc('initViewFileTreeSidebar', async (el: HTMLElement) => {
```

Comment:
Why it needs to use `registerGlobalInitFunc`? I don't see the "file tree" would be reloaded

Reply:
I thought that `registerGlobalInitFunc` could be used for any scenario with initialization requirements. So, is it only the pages that are loaded asynchronously that need to use `registerGlobalInitFunc`? If a non - asynchronously loaded page uses `registerGlobalInitFunc`, what problems might occur?

Reply:
For most cases, it is fine (somewhat abuse)

But for this case, it makes the problem more complex since you do other init works in `initViewFileTreeSidebar`.

If you could drop `registerGlobalInitFunc` here, the code will be clearer and we could clearly know it won't init more than once.